// testcafe/lib/test-run/index.js
'use strict';

exports.__esModule = true;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _create = require('babel-runtime/core-js/object/create');

var _create2 = _interopRequireDefault(_create);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _readFileRelative = require('read-file-relative');

var _promisifyEvent = require('promisify-event');

var _promisifyEvent2 = _interopRequireDefault(_promisifyEvent);

var _pinkie = require('pinkie');

var _pinkie2 = _interopRequireDefault(_pinkie);

var _mustache = require('mustache');

var _mustache2 = _interopRequireDefault(_mustache);

var _debugLogger = require('../notifications/debug-logger');

var _debugLogger2 = _interopRequireDefault(_debugLogger);

var _testcafeHammerhead = require('testcafe-hammerhead');

var _debugLog = require('./debug-log');

var _debugLog2 = _interopRequireDefault(_debugLog);

var _formattableAdapter = require('../errors/test-run/formattable-adapter');

var _formattableAdapter2 = _interopRequireDefault(_formattableAdapter);

var _errorList = require('../errors/error-list');

var _errorList2 = _interopRequireDefault(_errorList);

var _testRun = require('../errors/test-run/');

var _browserManipulationQueue = require('./browser-manipulation-queue');

var _browserManipulationQueue2 = _interopRequireDefault(_browserManipulationQueue);

var _clientMessages = require('./client-messages');

var _clientMessages2 = _interopRequireDefault(_clientMessages);

var _phase = require('./phase');

var _phase2 = _interopRequireDefault(_phase);

var _type = require('./commands/type');

var _type2 = _interopRequireDefault(_type);

var _executor = require('../assertions/executor');

var _executor2 = _interopRequireDefault(_executor);

var _delay = require('../utils/delay');

var _delay2 = _interopRequireDefault(_delay);

var _markerSymbol = require('./marker-symbol');

var _markerSymbol2 = _interopRequireDefault(_markerSymbol);

var _testRunTracker = require('../api/test-run-tracker');

var _testRunTracker2 = _interopRequireDefault(_testRunTracker);

var _phase3 = require('../role/phase');

var _phase4 = _interopRequireDefault(_phase3);

var _bookmark = require('./bookmark');

var _bookmark2 = _interopRequireDefault(_bookmark);

var _clientFunctionBuilder = require('../client-functions/client-function-builder');

var _clientFunctionBuilder2 = _interopRequireDefault(_clientFunctionBuilder);

var _browserManipulation = require('./commands/browser-manipulation');

var _actions = require('./commands/actions');

var _service = require('./commands/service');

var _utils = require('./commands/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

//Const
var TEST_RUN_TEMPLATE = (0, _readFileRelative.readSync)('../client/test-run/index.js.mustache');
var IFRAME_TEST_RUN_TEMPLATE = (0, _readFileRelative.readSync)('../client/test-run/iframe.js.mustache');
var TEST_DONE_CONFIRMATION_RESPONSE = 'test-done-confirmation';
var MAX_RESPONSE_DELAY = 2 * 60 * 1000;

var TestRun = function (_Session) {
    (0, _inherits3.default)(TestRun, _Session);

    function TestRun(test, browserConnection, screenshotCapturer, warningLog, opts) {
        (0, _classCallCheck3.default)(this, TestRun);

        var uploadsRoot = _path2.default.dirname(test.fixture.path);

        var _this = (0, _possibleConstructorReturn3.default)(this, _Session.call(this, uploadsRoot));

        _this[_markerSymbol2.default] = true;

        _this.opts = opts;
        _this.test = test;
        _this.browserConnection = browserConnection;

        _this.phase = _phase2.default.initial;

        _this.driverTaskQueue = [];
        _this.testDoneCommandQueued = false;

        _this.activeDialogHandler = null;
        _this.activeIframeSelector = null;
        _this.speed = _this.opts.speed;

        _this.pendingRequest = null;
        _this.pendingPageError = null;

        _this.controller = null;
        _this.ctx = (0, _create2.default)(null);
        _this.fixtureCtx = null;

        _this.currentRoleId = null;
        _this.usedRoleStates = (0, _create2.default)(null);

        _this.errs = [];

        _this.lastDriverStatusId = null;
        _this.lastDriverStatusResponse = null;

        _this.fileDownloadingHandled = false;
        _this.resolveWaitForFileDownloadingPromise = null;

        _this.debugging = _this.opts.debugMode;
        _this.disableDebugBreakpoints = false;

        _this.browserManipulationQueue = new _browserManipulationQueue2.default(browserConnection, screenshotCapturer, warningLog);

        _this.debugLog = new _debugLog2.default(_this.browserConnection.userAgent);

        _this.injectable.scripts.push('/testcafe-core.js');
        _this.injectable.scripts.push('/testcafe-ui.js');
        _this.injectable.scripts.push('/testcafe-automation.js');
        _this.injectable.scripts.push('/testcafe-driver.js');
        _this.injectable.styles.push('/testcafe-ui-styles.css');
        return _this;
    }

    // Hammerhead payload


    TestRun.prototype._getPayloadScript = function _getPayloadScript() {
        this.fileDownloadingHandled = false;
        this.resolveWaitForFileDownloadingPromise = null;

        return _mustache2.default.render(TEST_RUN_TEMPLATE, {
            testRunId: (0, _stringify2.default)(this.id),
            browserId: (0, _stringify2.default)(this.browserConnection.id),
            browserHeartbeatUrl: (0, _stringify2.default)(this.browserConnection.heartbeatUrl),
            browserStatusUrl: (0, _stringify2.default)(this.browserConnection.statusUrl),
            userAgent: (0, _stringify2.default)(this.browserConnection.userAgent),
            testName: (0, _stringify2.default)(this.test.name),
            fixtureName: (0, _stringify2.default)(this.test.fixture.name),
            selectorTimeout: this.opts.selectorTimeout,
            skipJsErrors: this.opts.skipJsErrors,
            speed: this.speed,
            dialogHandler: (0, _stringify2.default)(this.activeDialogHandler)
        });
    };

    TestRun.prototype._getIframePayloadScript = function _getIframePayloadScript() {
        return _mustache2.default.render(IFRAME_TEST_RUN_TEMPLATE, {
            testRunId: (0, _stringify2.default)(this.id),
            selectorTimeout: this.opts.selectorTimeout,
            speed: this.speed,
            dialogHandler: (0, _stringify2.default)(this.activeDialogHandler)
        });
    };

    // Hammerhead handlers


    TestRun.prototype.getAuthCredentials = function getAuthCredentials() {
        return this.test.authCredentials;
    };

    TestRun.prototype.handleFileDownload = function handleFileDownload() {
        if (this.resolveWaitForFileDownloadingPromise) {
            this.resolveWaitForFileDownloadingPromise(true);
            this.resolveWaitForFileDownloadingPromise = null;
        } else this.fileDownloadingHandled = true;
    };

    TestRun.prototype.handlePageError = function handlePageError(ctx, err) {
        this.pendingPageError = new _testRun.PageLoadError(err);

        ctx.redirect(ctx.toProxyUrl('about:error'));
    };

    // Test function execution


    TestRun.prototype._executeTestFn = function () {
        var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(phase, fn) {
            var screenshotPath;
            return _regenerator2.default.wrap(function _callee$(_context) {
                while (1) {
                    switch (_context.prev = _context.next) {
                        case 0:
                            this.phase = phase;

                            _context.prev = 1;
                            _context.next = 4;
                            return fn(this);

                        case 4:
                            _context.next = 15;
                            break;

                        case 6:
                            _context.prev = 6;
                            _context.t0 = _context['catch'](1);
                            screenshotPath = null;

                            if (!this.opts.takeScreenshotsOnFails) {
                                _context.next = 13;
                                break;
                            }

                            _context.next = 12;
                            return this.executeCommand(new _browserManipulation.TakeScreenshotOnFailCommand());

                        case 12:
                            screenshotPath = _context.sent;

                        case 13:

                            this.addError(_context.t0, screenshotPath);
                            return _context.abrupt('return', false);

                        case 15:
                            return _context.abrupt('return', !this._addPendingPageErrorIfAny());

                        case 16:
                        case 'end':
                            return _context.stop();
                    }
                }
            }, _callee, this, [[1, 6]]);
        }));

        function _executeTestFn(_x, _x2) {
            return _ref.apply(this, arguments);
        }

        return _executeTestFn;
    }();

    TestRun.prototype._runBeforeHook = function () {
        var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2() {
            return _regenerator2.default.wrap(function _callee2$(_context2) {
                while (1) {
                    switch (_context2.prev = _context2.next) {
                        case 0:
                            if (!this.test.beforeFn) {
                                _context2.next = 4;
                                break;
                            }

                            _context2.next = 3;
                            return this._executeTestFn(_phase2.default.inTestBeforeHook, this.test.beforeFn);

                        case 3:
                            return _context2.abrupt('return', _context2.sent);

                        case 4:
                            if (!this.test.fixture.beforeEachFn) {
                                _context2.next = 8;
                                break;
                            }

                            _context2.next = 7;
                            return this._executeTestFn(_phase2.default.inFixtureBeforeEachHook, this.test.fixture.beforeEachFn);

                        case 7:
                            return _context2.abrupt('return', _context2.sent);

                        case 8:
                            return _context2.abrupt('return', true);

                        case 9:
                        case 'end':
                            return _context2.stop();
                    }
                }
            }, _callee2, this);
        }));

        function _runBeforeHook() {
            return _ref2.apply(this, arguments);
        }

        return _runBeforeHook;
    }();

    TestRun.prototype._runAfterHook = function () {
        var _ref3 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3() {
            return _regenerator2.default.wrap(function _callee3$(_context3) {
                while (1) {
                    switch (_context3.prev = _context3.next) {
                        case 0:
                            if (!this.test.afterFn) {
                                _context3.next = 4;
                                break;
                            }

                            _context3.next = 3;
                            return this._executeTestFn(_phase2.default.inTestAfterHook, this.test.afterFn);

                        case 3:
                            return _context3.abrupt('return', _context3.sent);

                        case 4:
                            if (!this.test.fixture.afterEachFn) {
                                _context3.next = 8;
                                break;
                            }

                            _context3.next = 7;
                            return this._executeTestFn(_phase2.default.inFixtureAfterEachHook, this.test.fixture.afterEachFn);

                        case 7:
                            return _context3.abrupt('return', _context3.sent);

                        case 8:
                            return _context3.abrupt('return', true);

                        case 9:
                        case 'end':
                            return _context3.stop();
                    }
                }
            }, _callee3, this);
        }));

        function _runAfterHook() {
            return _ref3.apply(this, arguments);
        }

        return _runAfterHook;
    }();

    TestRun.prototype.start = function () {
        var _ref4 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee4() {
            return _regenerator2.default.wrap(function _callee4$(_context4) {
                while (1) {
                    switch (_context4.prev = _context4.next) {
                        case 0:
                            _testRunTracker2.default.activeTestRuns[this.id] = this;

                            this.emit('start');

                            _context4.next = 4;
                            return this._runBeforeHook();

                        case 4:
                            if (!_context4.sent) {
                                _context4.next = 9;
                                break;
                            }

                            _context4.next = 7;
                            return this._executeTestFn(_phase2.default.inTest, this.test.fn);

                        case 7:
                            _context4.next = 9;
                            return this._runAfterHook();

                        case 9:
                            _context4.next = 11;
                            return this.executeCommand(new _service.TestDoneCommand());

                        case 11:
                            this._addPendingPageErrorIfAny();

                            delete _testRunTracker2.default.activeTestRuns[this.id];

                            this.emit('done');

                        case 14:
                        case 'end':
                            return _context4.stop();
                    }
                }
            }, _callee4, this);
        }));

        function start() {
            return _ref4.apply(this, arguments);
        }

        return start;
    }();

    // Errors


    TestRun.prototype._addPendingPageErrorIfAny = function _addPendingPageErrorIfAny() {
        if (this.pendingPageError) {
            this.addError(this.pendingPageError);
            this.pendingPageError = null;
            return true;
        }

        return false;
    };

    TestRun.prototype.addError = function addError(err, screenshotPath) {
        var _this2 = this;

        var errList = err instanceof _errorList2.default ? err.items : [err];

        errList.forEach(function (item) {
            var adapter = new _formattableAdapter2.default(item, {
                userAgent: _this2.browserConnection.userAgent,
                screenshotPath: screenshotPath || '',
                testRunPhase: _this2.phase
            });

            _this2.errs.push(adapter);
        });
    };

    // Task queue


    TestRun.prototype._enqueueCommand = function _enqueueCommand(command, callsite) {
        var _this3 = this;

        if (this.pendingRequest) this._resolvePendingRequest(command);

        return new _pinkie2.default(function (resolve, reject) {
            return _this3.driverTaskQueue.push({ command: command, resolve: resolve, reject: reject, callsite: callsite });
        });
    };

    TestRun.prototype._enqueueBrowserManipulation = function _enqueueBrowserManipulation(command, callsite) {
        this.browserManipulationQueue.push(command);
        return this.executeCommand(new _service.PrepareBrowserManipulationCommand(command.type), callsite);
    };

    TestRun.prototype._enqueueSetBreakpointCommand = function () {
        var _ref5 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee5(callsite) {
            return _regenerator2.default.wrap(function _callee5$(_context5) {
                while (1) {
                    switch (_context5.prev = _context5.next) {
                        case 0:
                            _debugLogger2.default.showBreakpoint(this.id, this.browserConnection.userAgent, callsite);

                            _context5.next = 3;
                            return this._enqueueCommand(new _service.SetBreakpointCommand(), callsite);

                        case 3:
                            this.debugging = _context5.sent;

                        case 4:
                        case 'end':
                            return _context5.stop();
                    }
                }
            }, _callee5, this);
        }));

        function _enqueueSetBreakpointCommand(_x3) {
            return _ref5.apply(this, arguments);
        }

        return _enqueueSetBreakpointCommand;
    }();

    TestRun.prototype._removeAllNonServiceTasks = function _removeAllNonServiceTasks() {
        this.driverTaskQueue = this.driverTaskQueue.filter(function (driverTask) {
            return (0, _utils.isServiceCommand)(driverTask.command);
        });

        this.browserManipulationQueue.removeAllNonServiceManipulations();
    };

    // Current driver task


    TestRun.prototype._resolveCurrentDriverTask = function _resolveCurrentDriverTask(result) {
        this.currentDriverTask.resolve(result);
        this.driverTaskQueue.shift();

        if (this.testDoneCommandQueued) this._removeAllNonServiceTasks();
    };

    TestRun.prototype._rejectCurrentDriverTask = function _rejectCurrentDriverTask(err) {
        err.callsite = err.callsite || this.driverTaskQueue[0].callsite;

        this.currentDriverTask.reject(err);
        this._removeAllNonServiceTasks();
    };

    // Pending request


    TestRun.prototype._clearPendingRequest = function _clearPendingRequest() {
        if (this.pendingRequest) {
            clearTimeout(this.pendingRequest.responseTimeout);
            this.pendingRequest = null;
        }
    };

    TestRun.prototype._resolvePendingRequest = function _resolvePendingRequest(command) {
        this.lastDriverStatusResponse = command;
        this.pendingRequest.resolve(command);
        this._clearPendingRequest();
    };

    // Handle driver request


    TestRun.prototype._fulfillCurrentDriverTask = function _fulfillCurrentDriverTask(driverStatus) {
        if (driverStatus.executionError) this._rejectCurrentDriverTask(driverStatus.executionError);else this._resolveCurrentDriverTask(driverStatus.result);
    };

    TestRun.prototype._handlePageErrorStatus = function _handlePageErrorStatus(pageError) {
        if (this.currentDriverTask && (0, _utils.isCommandRejectableByPageError)(this.currentDriverTask.command)) {
            this._rejectCurrentDriverTask(pageError);
            this.pendingPageError = null;

            return true;
        }

        this.pendingPageError = this.pendingPageError || pageError;

        return false;
    };

    TestRun.prototype._handleDriverRequest = function _handleDriverRequest(driverStatus) {
        var pageError = this.pendingPageError || driverStatus.pageError;

        var currentTaskRejectedByError = pageError && this._handlePageErrorStatus(pageError);

        if (!currentTaskRejectedByError && driverStatus.isCommandResult) {
            if (this.currentDriverTask.command.type === _type2.default.testDone) {
                this._resolveCurrentDriverTask();

                return TEST_DONE_CONFIRMATION_RESPONSE;
            }

            this._fulfillCurrentDriverTask(driverStatus);
        }

        return this.currentDriverTask ? this.currentDriverTask.command : null;
    };

    // Execute command


    TestRun.prototype._executeAssertion = function () {
        var _ref6 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee6(command, callsite) {
            var _this4 = this;

            var executor;
            return _regenerator2.default.wrap(function _callee6$(_context6) {
                while (1) {
                    switch (_context6.prev = _context6.next) {
                        case 0:
                            executor = new _executor2.default(command, callsite);


                            executor.once('start-assertion-retries', function (timeout) {
                                return _this4._enqueueCommand(new _service.ShowAssertionRetriesStatusCommand(timeout));
                            });
                            executor.once('end-assertion-retries', function (success) {
                                return _this4._enqueueCommand(new _service.HideAssertionRetriesStatusCommand(success));
                            });

                            return _context6.abrupt('return', executor.run());

                        case 4:
                        case 'end':
                            return _context6.stop();
                    }
                }
            }, _callee6, this);
        }));

        function _executeAssertion(_x4, _x5) {
            return _ref6.apply(this, arguments);
        }

        return _executeAssertion;
    }();

    TestRun.prototype._adjustConfigurationWithCommand = function _adjustConfigurationWithCommand(command) {
        if (command.type === _type2.default.testDone) {
            this.testDoneCommandQueued = true;
            _debugLogger2.default.hideBreakpoint(this.id);
        } else if (command.type === _type2.default.setNativeDialogHandler) this.activeDialogHandler = command.dialogHandler;else if (command.type === _type2.default.switchToIframe) this.activeIframeSelector = command.selector;else if (command.type === _type2.default.switchToMainWindow) this.activeIframeSelector = null;else if (command.type === _type2.default.setTestSpeed) this.speed = command.speed;else if (command.type === _type2.default.debug) this.debugging = true;
    };

    TestRun.prototype._setBreakpointIfNecessary = function () {
        var _ref7 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee7(command, callsite) {
            return _regenerator2.default.wrap(function _callee7$(_context7) {
                while (1) {
                    switch (_context7.prev = _context7.next) {
                        case 0:
                            if (!(!this.disableDebugBreakpoints && this.debugging && (0, _utils.canSetDebuggerBreakpointBeforeCommand)(command))) {
                                _context7.next = 3;
                                break;
                            }

                            _context7.next = 3;
                            return this._enqueueSetBreakpointCommand(callsite);

                        case 3:
                        case 'end':
                            return _context7.stop();
                    }
                }
            }, _callee7, this);
        }));

        function _setBreakpointIfNecessary(_x6, _x7) {
            return _ref7.apply(this, arguments);
        }

        return _setBreakpointIfNecessary;
    }();

    TestRun.prototype.executeCommand = function () {
        var _ref8 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee8(command, callsite) {
            return _regenerator2.default.wrap(function _callee8$(_context8) {
                while (1) {
                    switch (_context8.prev = _context8.next) {
                        case 0:
                            this.debugLog.command(command);

                            if (!(this.pendingPageError && (0, _utils.isCommandRejectableByPageError)(command))) {
                                _context8.next = 3;
                                break;
                            }

                            return _context8.abrupt('return', this._rejectCommandWithPageError(callsite));

                        case 3:

                            this._adjustConfigurationWithCommand(command);

                            _context8.next = 6;
                            return this._setBreakpointIfNecessary(command, callsite);

                        case 6:
                            if (!(0, _utils.isBrowserManipulationCommand)(command)) {
                                _context8.next = 8;
                                break;
                            }

                            return _context8.abrupt('return', this._enqueueBrowserManipulation(command, callsite));

                        case 8:
                            if (!(command.type === _type2.default.wait)) {
                                _context8.next = 10;
                                break;
                            }

                            return _context8.abrupt('return', (0, _delay2.default)(command.timeout));

                        case 10:
                            if (!(command.type === _type2.default.debug)) {
                                _context8.next = 14;
                                break;
                            }

                            _context8.next = 13;
                            return this._enqueueSetBreakpointCommand(callsite);

                        case 13:
                            return _context8.abrupt('return', _context8.sent);

                        case 14:
                            if (!(command.type === _type2.default.useRole)) {
                                _context8.next = 18;
                                break;
                            }

                            _context8.next = 17;
                            return this._useRole(command.role, callsite);

                        case 17:
                            return _context8.abrupt('return', _context8.sent);

                        case 18:
                            if (!(command.type === _type2.default.assertion)) {
                                _context8.next = 20;
                                break;
                            }

                            return _context8.abrupt('return', this._executeAssertion(command, callsite));

                        case 20:
                            return _context8.abrupt('return', this._enqueueCommand(command, callsite));

                        case 21:
                        case 'end':
                            return _context8.stop();
                    }
                }
            }, _callee8, this);
        }));

        function executeCommand(_x8, _x9) {
            return _ref8.apply(this, arguments);
        }

        return executeCommand;
    }();

    TestRun.prototype._rejectCommandWithPageError = function _rejectCommandWithPageError(callsite) {
        var err = this.pendingPageError;

        err.callsite = callsite;
        this.pendingPageError = null;

        return _pinkie2.default.reject(err);
    };

    // Role management


    TestRun.prototype.switchToCleanRun = function () {
        var _ref9 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee9() {
            var removeDialogHandlerCommand, setSpeedCommand;
            return _regenerator2.default.wrap(function _callee9$(_context9) {
                while (1) {
                    switch (_context9.prev = _context9.next) {
                        case 0:
                            this.ctx = (0, _create2.default)(null);
                            this.fixtureCtx = (0, _create2.default)(null);

                            this.useStateSnapshot(null);

                            if (!this.activeDialogHandler) {
                                _context9.next = 7;
                                break;
                            }

                            removeDialogHandlerCommand = new _actions.SetNativeDialogHandlerCommand({ dialogHandler: { fn: null } });
                            _context9.next = 7;
                            return this.executeCommand(removeDialogHandlerCommand);

                        case 7:
                            if (!(this.speed !== this.opts.speed)) {
                                _context9.next = 11;
                                break;
                            }

                            setSpeedCommand = new _actions.SetTestSpeedCommand({ speed: this.opts.speed });
                            _context9.next = 11;
                            return this.executeCommand(setSpeedCommand);

                        case 11:
                        case 'end':
                            return _context9.stop();
                    }
                }
            }, _callee9, this);
        }));

        function switchToCleanRun() {
            return _ref9.apply(this, arguments);
        }

        return switchToCleanRun;
    }();

    TestRun.prototype._getStateSnapshotFromRole = function () {
        var _ref10 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee10(role) {
            var prevPhase;
            return _regenerator2.default.wrap(function _callee10$(_context10) {
                while (1) {
                    switch (_context10.prev = _context10.next) {
                        case 0:
                            prevPhase = this.phase;


                            this.phase = _phase2.default.inRoleInitializer;

                            if (!(role.phase === _phase4.default.uninitialized)) {
                                _context10.next = 7;
                                break;
                            }

                            _context10.next = 5;
                            return role.initialize(this);

                        case 5:
                            _context10.next = 10;
                            break;

                        case 7:
                            if (!(role.phase === _phase4.default.pendingInitialization)) {
                                _context10.next = 10;
                                break;
                            }

                            _context10.next = 10;
                            return (0, _promisifyEvent2.default)(role, 'initialized');

                        case 10:
                            if (!role.initErr) {
                                _context10.next = 12;
                                break;
                            }

                            throw role.initErr;

                        case 12:

                            this.phase = prevPhase;

                            return _context10.abrupt('return', role.stateSnapshot);

                        case 14:
                        case 'end':
                            return _context10.stop();
                    }
                }
            }, _callee10, this);
        }));

        function _getStateSnapshotFromRole(_x10) {
            return _ref10.apply(this, arguments);
        }

        return _getStateSnapshotFromRole;
    }();

    TestRun.prototype._useRole = function () {
        var _ref11 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee11(role, callsite) {
            var bookmark, stateSnapshot;
            return _regenerator2.default.wrap(function _callee11$(_context11) {
                while (1) {
                    switch (_context11.prev = _context11.next) {
                        case 0:
                            if (!(this.phase === _phase2.default.inRoleInitializer)) {
                                _context11.next = 2;
                                break;
                            }

                            throw new _testRun.RoleSwitchInRoleInitializerError(callsite);

                        case 2:

                            this.disableDebugBreakpoints = true;

                            bookmark = new _bookmark2.default(this, role);
                            _context11.next = 6;
                            return bookmark.init();

                        case 6:

                            if (this.currentRoleId) this.usedRoleStates[this.currentRoleId] = this.getStateSnapshot();

                            _context11.t0 = this.usedRoleStates[role.id];

                            if (_context11.t0) {
                                _context11.next = 12;
                                break;
                            }

                            _context11.next = 11;
                            return this._getStateSnapshotFromRole(role);

                        case 11:
                            _context11.t0 = _context11.sent;

                        case 12:
                            stateSnapshot = _context11.t0;


                            this.useStateSnapshot(stateSnapshot);

                            this.currentRoleId = role.id;

                            _context11.next = 17;
                            return bookmark.restore(callsite);

                        case 17:

                            this.disableDebugBreakpoints = false;

                        case 18:
                        case 'end':
                            return _context11.stop();
                    }
                }
            }, _callee11, this);
        }));

        function _useRole(_x11, _x12) {
            return _ref11.apply(this, arguments);
        }

        return _useRole;
    }();

    // Get current URL


    TestRun.prototype.getCurrentUrl = function () {
        var _ref12 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee12() {
            var builder, getLocation;
            return _regenerator2.default.wrap(function _callee12$(_context12) {
                while (1) {
                    switch (_context12.prev = _context12.next) {
                        case 0:
                            builder = new _clientFunctionBuilder2.default(function () {
                                /* eslint-disable no-undef */
                                return window.location.href;
                                /* eslint-enable no-undef */
                            }, { boundTestRun: this });
                            getLocation = builder.getFunction();
                            _context12.next = 4;
                            return getLocation();

                        case 4:
                            return _context12.abrupt('return', _context12.sent);

                        case 5:
                        case 'end':
                            return _context12.stop();
                    }
                }
            }, _callee12, this);
        }));

        function getCurrentUrl() {
            return _ref12.apply(this, arguments);
        }

        return getCurrentUrl;
    }();

    (0, _createClass3.default)(TestRun, [{
        key: 'currentDriverTask',
        get: function get() {
            return this.driverTaskQueue[0];
        }
    }]);
    return TestRun;
}(_testcafeHammerhead.Session);

// Service message handlers


exports.default = TestRun;
var ServiceMessages = TestRun.prototype;

ServiceMessages[_clientMessages2.default.ready] = function (msg) {
    var _this5 = this;

    this.debugLog.driverMessage(msg);

    this._clearPendingRequest();

    // NOTE: the driver sends the status for the second time if it didn't get a response at the
    // first try. This is possible when the page was unloaded after the driver sent the status.
    if (msg.status.id === this.lastDriverStatusId) return this.lastDriverStatusResponse;

    this.lastDriverStatusId = msg.status.id;
    this.lastDriverStatusResponse = this._handleDriverRequest(msg.status);

    if (this.lastDriverStatusResponse) return this.lastDriverStatusResponse;

    // NOTE: browsers abort an opened xhr request after a certain timeout (the actual duration depends on the browser).
    // To avoid this, we send an empty response after 2 minutes if we didn't get any command.
    var responseTimeout = setTimeout(function () {
        return _this5._resolvePendingRequest(null);
    }, MAX_RESPONSE_DELAY);

    return new _pinkie2.default(function (resolve, reject) {
        _this5.pendingRequest = { resolve: resolve, reject: reject, responseTimeout: responseTimeout };
    });
};

ServiceMessages[_clientMessages2.default.readyForBrowserManipulation] = function () {
    var _ref13 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee13(msg) {
        return _regenerator2.default.wrap(function _callee13$(_context13) {
            while (1) {
                switch (_context13.prev = _context13.next) {
                    case 0:
                        this.debugLog.driverMessage(msg);

                        _context13.next = 3;
                        return this.browserManipulationQueue.executePendingManipulation(msg);

                    case 3:
                        return _context13.abrupt('return', _context13.sent);

                    case 4:
                    case 'end':
                        return _context13.stop();
                }
            }
        }, _callee13, this);
    }));

    return function (_x13) {
        return _ref13.apply(this, arguments);
    };
}();

ServiceMessages[_clientMessages2.default.waitForFileDownload] = function (msg) {
    var _this6 = this;

    this.debugLog.driverMessage(msg);

    return new _pinkie2.default(function (resolve) {
        if (_this6.fileDownloadingHandled) {
            _this6.fileDownloadingHandled = false;
            resolve(true);
        } else _this6.resolveWaitForFileDownloadingPromise = resolve;
    });
};
module.exports = exports['default'];
