// testcafe-hammerhead/lib/processing/script/transformers/location-get.js
'use strict';

exports.__esModule = true;

var _nodeBuilder = require('../node-builder');

var _instruction = require('../instruction');

var _instruction2 = _interopRequireDefault(_instruction);

var _esotope = require('../tools/esotope');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Transform:
// location -->
// __get$Loc(location)

exports.default = {
    nodeReplacementRequireTransform: false,

    nodeTypes: [_esotope.Syntax.Identifier],

    condition: function condition(node, parent) {
        if (node.name !== 'location') return false;

        // Skip: var location = value;
        if (parent.type === _esotope.Syntax.VariableDeclarator && parent.id === node) return false;

        // Skip: location = value
        if (parent.type === _esotope.Syntax.AssignmentExpression && parent.left === node) return false;

        // Skip: function location() {}
        if (parent.type === _esotope.Syntax.FunctionDeclaration && parent.id === node) return false;

        // Skip: object.location || location.field
        if (parent.type === _esotope.Syntax.MemberExpression) return false;

        // Skip: { location: value }
        if (parent.type === _esotope.Syntax.Property && parent.key === node) return false;

        // Skip: location++ || location-- || ++location || --location
        if (parent.type === _esotope.Syntax.UpdateExpression && parent.operator === '++' || parent.operator === '--') return false;

        // Skip: function (location) { ... } || function func(location) { ... }
        if ((parent.type === _esotope.Syntax.FunctionExpression || parent.type === _esotope.Syntax.FunctionDeclaration) && parent.params.indexOf(node) !== -1) return false;

        // Skip already transformed: __get$Loc(location)
        if (parent.type === _esotope.Syntax.CallExpression && parent.callee.name === _instruction2.default.getLocation) return false;

        return true;
    },

    run: function run() {
        return (0, _nodeBuilder.createLocationGetWrapper)();
    }
}; // -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------

module.exports = exports['default'];
