// testcafe-hammerhead/lib/upload/form-data-entry.js
'use strict';

exports.__esModule = true;

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _buffer = require('../utils/buffer');

var bufferUtils = _interopRequireWildcard(_buffer);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Const
var INPUT_NAME_RE = /;\s*name="([^"]*)"/i;
var FILE_NAME_RE = /;\s*filename="([^"]*)"/i;
var HEADER_RE = /^(.+?):\s*(.*)$/;

// FormDataEntry

var FormDataEntry = function () {
    function FormDataEntry() {
        (0, _classCallCheck3.default)(this, FormDataEntry);

        this.body = [];
        this.headers = {};
        this.name = null;
        this.fileName = null;
    }

    FormDataEntry.prototype._parseContentDisposition = function _parseContentDisposition(contentDisposition) {
        var inputNameMatch = contentDisposition.match(INPUT_NAME_RE);
        var fileNameMatch = contentDisposition.match(FILE_NAME_RE);

        this.name = inputNameMatch && inputNameMatch[1];
        this.fileName = fileNameMatch && fileNameMatch[1];
    };

    FormDataEntry.prototype._setContentDisposition = function _setContentDisposition(name, fileName) {
        this.name = name;
        this.fileName = fileName;

        this.headers['Content-Disposition'] = 'form-data; name="' + name + '"; filename="' + fileName + '"';
    };

    // API


    FormDataEntry.prototype.addFileInfo = function addFileInfo(fileInfo, idx) {
        var file = fileInfo.files[idx];

        this._setContentDisposition(fileInfo.name, file.name);

        this.body = [new Buffer(file.data, 'base64')];
        this.headers['Content-Type'] = file.type;
    };

    FormDataEntry.prototype.setHeader = function setHeader(header, newValue) {
        var headerMatch = header.match(HEADER_RE);
        var name = headerMatch[1];
        var value = newValue || headerMatch[2];

        this.headers[name] = value;

        if (name === 'Content-Disposition') this._parseContentDisposition(value);
    };

    FormDataEntry.prototype.toBuffer = function toBuffer() {
        var _this = this;

        var chunks = [];

        (0, _keys2.default)(this.headers).forEach(function (name) {
            var value = _this.headers[name];

            chunks.push(new Buffer(name + ': ' + value));
            chunks.push(bufferUtils.CRLF);
        });

        chunks.push(bufferUtils.CRLF);
        chunks = chunks.concat(this.body);

        return Buffer.concat(chunks);
    };

    return FormDataEntry;
}();

exports.default = FormDataEntry;
module.exports = exports['default'];
