// testcafe-browser-tools/lib/aliases.js
'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default').default;

exports.__esModule = true;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _binaries = require('./binaries');

var _binaries2 = _interopRequireDefault(_binaries);

var ALIASES = {
    'ie': {
        nameRe: /ie|internet explorer/i,
        cmd: ''
    },

    'firefox': {
        nameRe: /firefox|mozilla/i,
        cmd: '-override "' + _path2.default.join(__dirname, '../data/override.ini') + '" -new-window',
        macOpenCmdTemplate: 'open -a "{{{path}}}" {{{pageUrl}}} --args {{{cmd}}}',
        linuxBinaries: ['firefox']
    },

    'chrome': {
        nameRe: /chrome/i,
        cmd: '--no-first-run --new-window',
        macOpenCmdTemplate: 'open -n -a "{{{path}}}" --args {{{pageUrl}}} {{{cmd}}}',
        linuxBinaries: ['google-chrome', 'google-chrome-stable', 'google-chrome-canary']
    },

    'chromium': {
        nameRe: /chromium/i,
        cmd: '--no-first-run --new-window',
        macOpenCmdTemplate: 'open -n -a "{{{path}}}" --args {{{pageUrl}}} {{{cmd}}}',
        linuxBinaries: ['chromium-browser', 'chromium']
    },

    'opera': {
        nameRe: /opera/i,
        cmd: '--new-window',
        macOpenCmdTemplate: 'open -n -a "{{{path}}}" --args {{{pageUrl}}} {{{cmd}}}',
        linuxBinaries: ['opera']
    },

    'safari': {
        nameRe: /safari/i,
        cmd: '',
        path: _binaries2.default.open,
        macOpenCmdTemplate: '/usr/bin/osascript "{{{path}}}" {{{pageUrl}}} --args {{{cmd}}}'
    },

    'edge': {
        nameRe: /edge/i,
        cmd: '',
        winOpenCmdTemplate: 'start microsoft-edge:"{{{pageUrl}}}"'
    }
};

exports.default = ALIASES;
module.exports = exports.default;
