// testcafe-browser-tools/lib/binaries.js
'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default').default;

exports.__esModule = true;

var _osFamily = require('os-family');

var _osFamily2 = _interopRequireDefault(_osFamily);

var _readFileRelative = require('read-file-relative');

var _utilsArch = require('./utils/arch');

var _utilsArch2 = _interopRequireDefault(_utilsArch);

var BINARIES = void 0;

if (_osFamily2.default.win) {
    BINARIES = {
        findWindow: _readFileRelative.toAbsPath('../bin/win/find-window.exe'),
        getWindowSize: _readFileRelative.toAbsPath('../bin/win/get-window-size.exe'),
        close: _readFileRelative.toAbsPath('../bin/win/close.exe'),
        screenshot: _readFileRelative.toAbsPath('../bin/win/screenshot.exe'),
        generateThumbnail: _readFileRelative.toAbsPath('../bin/win/generate-thumbnail.exe'),
        resize: _readFileRelative.toAbsPath('../bin/win/resize.exe'),
        maximize: _readFileRelative.toAbsPath('../bin/win/maximize.exe')
    };
} else if (_osFamily2.default.mac) {
    BINARIES = {
        open: _readFileRelative.toAbsPath('../bin/mac/open.scpt'),
        findWindow: _readFileRelative.toAbsPath('../bin/mac/find-window.scpt'),
        getWindowSize: _readFileRelative.toAbsPath('../bin/mac/get-window-size.scpt'),
        getWindowBounds: _readFileRelative.toAbsPath('../bin/mac/get-window-bounds.scpt'),
        getWindowMaxBounds: _readFileRelative.toAbsPath('../bin/mac/get-window-max-bounds'),
        setWindowBounds: _readFileRelative.toAbsPath('../bin/mac/set-window-bounds.scpt'),
        close: _readFileRelative.toAbsPath('../bin/mac/close.scpt'),
        screenshot: _readFileRelative.toAbsPath('../bin/mac/screenshot'),
        resize: _readFileRelative.toAbsPath('../bin/mac/resize.scpt'),
        generateThumbnail: _readFileRelative.toAbsPath('../bin/mac/generate-thumbnail')
    };
} else if (_osFamily2.default.linux) {
    BINARIES = {
        findWindow: _readFileRelative.toAbsPath('../bin/linux/find-window.sh'),
        close: _readFileRelative.toAbsPath('../bin/linux/' + _utilsArch2.default.bits + '/close'),
        getWindowSize: _readFileRelative.toAbsPath('../bin/linux/' + _utilsArch2.default.bits + '/get-window-size'),
        resize: _readFileRelative.toAbsPath('../bin/linux/' + _utilsArch2.default.bits + '/resize'),
        maximize: _readFileRelative.toAbsPath('../bin/linux/' + _utilsArch2.default.bits + '/maximize'),
        screenshot: _readFileRelative.toAbsPath('../bin/linux/' + _utilsArch2.default.bits + '/screenshot'),
        generateThumbnail: _readFileRelative.toAbsPath('../bin/linux/' + _utilsArch2.default.bits + '/generate-thumbnail')
    };
} else BINARIES = {};

exports.default = BINARIES;
module.exports = exports.default;
