// testcafe-browser-tools/lib/index.js
'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default').default;

exports.__esModule = true;

var _apiGetInstallations = require('./api/get-installations');

var _apiGetInstallations2 = _interopRequireDefault(_apiGetInstallations);

var _apiFindWindow = require('./api/find-window');

var _apiFindWindow2 = _interopRequireDefault(_apiFindWindow);

var _apiOpen = require('./api/open');

var _apiOpen2 = _interopRequireDefault(_apiOpen);

var _apiClose = require('./api/close');

var _apiClose2 = _interopRequireDefault(_apiClose);

var _apiResize = require('./api/resize');

var _apiResize2 = _interopRequireDefault(_apiResize);

var _apiMaximize = require('./api/maximize');

var _apiMaximize2 = _interopRequireDefault(_apiMaximize);

var _apiIsMaximized = require('./api/is-maximized');

var _apiIsMaximized2 = _interopRequireDefault(_apiIsMaximized);

var _apiIsValidDeviceName = require('./api/is-valid-device-name');

var _apiIsValidDeviceName2 = _interopRequireDefault(_apiIsValidDeviceName);

var _apiScreenshot = require('./api/screenshot');

var _apiScreenshot2 = _interopRequireDefault(_apiScreenshot);

var _apiGenerateThumbnail = require('./api/generate-thumbnail');

var _apiGenerateThumbnail2 = _interopRequireDefault(_apiGenerateThumbnail);

var _apiGetBrowserInfo = require('./api/get-browser-info');

var _apiGetBrowserInfo2 = _interopRequireDefault(_apiGetBrowserInfo);

var _apiGetViewportSize = require('./api/get-viewport-size');

var _apiGetViewportSize2 = _interopRequireDefault(_apiGetViewportSize);

exports.default = {
    getInstallations: _apiGetInstallations2.default,
    getBrowserInfo: _apiGetBrowserInfo2.default,
    getViewportSize: _apiGetViewportSize2.default,
    findWindow: _apiFindWindow2.default,
    open: _apiOpen2.default,
    close: _apiClose2.default,
    resize: _apiResize2.default,
    maximize: _apiMaximize2.default,
    isMaximized: _apiIsMaximized2.default,
    screenshot: _apiScreenshot2.default,
    generateThumbnail: _apiGenerateThumbnail2.default,
    isValidDeviceName: _apiIsValidDeviceName2.default
};
module.exports = exports.default;
