// testcafe-hammerhead/lib/processing/script/transform.js
'use strict';

exports.__esModule = true;
exports.default = transform;

var _transformers = require('./transformers');

var _transformers2 = _interopRequireDefault(_transformers);

var _replaceNode = require('./transformers/replace-node');

var _replaceNode2 = _interopRequireDefault(_replaceNode);

var _esotope = require('./tools/esotope');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// NOTE: We should avoid using native object prototype methods,
// since they can be overriden by the client code. (GH-245)
var objectToString = Object.prototype.toString; // -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------

function getChange(node, parent, key) {
    return {
        start: node.originStart,
        end: node.originEnd,
        index: Array.isArray(parent[key]) ? parent[key].indexOf(node) : -1,

        parent: parent,
        key: key
    };
}

// NOTE: There is an issue with processing `new` expressions. `new a.src.b()` will be transformed
// to `new __get$(a, 'src').b()`, which is wrong. The correct result is `new (__get$(a, 'src')).b()`.
// To solve this problem, we add a 'state' entity. This entity stores the "new" expression, so that
// we can add it to the changes when the transformation is found.
function createState(currState, node, parent, key, hasTransformedAncestor) {
    var isNewExpression = node.type === _esotope.Syntax.NewExpression;
    var isNewExpressionAncestor = isNewExpression && !currState.newExpressionAncestor;

    return {
        hasTransformedAncestor: currState.hasTransformedAncestor || hasTransformedAncestor,
        newExpressionAncestor: isNewExpressionAncestor ? node : currState.newExpressionAncestor,
        newExpressionAncestorParent: isNewExpressionAncestor ? parent : currState.newExpressionAncestorParent,
        newExpressionAncestorKey: isNewExpressionAncestor ? key : currState.newExpressionAncestorKey
    };
}

function transformChildNodes(node, changes, state) {
    for (var key in node) {
        if (node.hasOwnProperty(key)) {
            var childNode = node[key];

            if (objectToString.call(childNode) === '[object Array]') {
                for (var j = 0; j < childNode.length; j++) {
                    transform(childNode[j], node, key, changes, state);
                }
            } else transform(childNode, node, key, changes, state);
        }
    }
}

function isNodeTransformed(node) {
    return node.originStart !== void 0 && node.originEnd !== void 0;
}

function addChangeForTransformedNode(state, changes, replacement, parent, key) {
    var hasTransformedAncestor = state.hasTransformedAncestor;

    hasTransformedAncestor |= state.newExpressionAncestor && isNodeTransformed(state.newExpressionAncestor);

    if (!hasTransformedAncestor) {
        if (state.newExpressionAncestor) {
            (0, _replaceNode2.default)(state.newExpressionAncestor, state.newExpressionAncestor, state.newExpressionAncestorParent, state.newExpressionAncestorKey);
            changes.push(getChange(state.newExpressionAncestor, state.newExpressionAncestorParent, state.newExpressionAncestorKey));
        } else changes.push(getChange(replacement, parent, key));
    }
}

function transform(node, parent, key, changes, state, reTransform) {
    state = state || {};
    changes = changes || [];

    if (!node || typeof node !== 'object') return null;

    var nodeChanged = false;

    if (isNodeTransformed(node) && !reTransform) {
        addChangeForTransformedNode(state, changes, node, parent, key);
        nodeChanged = true;
    } else {
        var nodeTransformers = _transformers2.default[node.type];

        if (nodeTransformers) {
            for (var i = 0; i < nodeTransformers.length; i++) {
                var transformer = nodeTransformers[i];

                if (transformer.condition(node, parent)) {
                    var replacement = transformer.run(node, parent, key);

                    if (replacement) {
                        (0, _replaceNode2.default)(node, replacement, parent, key);
                        nodeChanged = true;

                        addChangeForTransformedNode(state, changes, replacement, parent, key);

                        if (transformer.nodeReplacementRequireTransform) {
                            var newState = createState(state, replacement, parent, key, nodeChanged);

                            transform(replacement, parent, key, changes, newState, true);

                            return changes;
                        }

                        break;
                    }
                }
            }
        }
    }

    var childNodesState = createState(state, node, parent, key, nodeChanged);

    transformChildNodes(node, changes, childNodesState);

    return changes;
}
module.exports = exports['default'];
