// testcafe/lib/test-run/commands/service.js
'use strict';

exports.__esModule = true;
exports.TestDoneCommand = exports.SetBreakpointCommand = exports.HideAssertionRetriesStatusCommand = exports.ShowAssertionRetriesStatusCommand = exports.PrepareBrowserManipulationCommand = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _type = require('./type');

var _type2 = _interopRequireDefault(_type);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Commands
var PrepareBrowserManipulationCommand = exports.PrepareBrowserManipulationCommand = function PrepareBrowserManipulationCommand(manipulationCommandType) {
    (0, _classCallCheck3.default)(this, PrepareBrowserManipulationCommand);

    this.type = _type2.default.prepareBrowserManipulation;
    this.manipulationCommandType = manipulationCommandType;
};

var ShowAssertionRetriesStatusCommand = exports.ShowAssertionRetriesStatusCommand = function ShowAssertionRetriesStatusCommand(timeout) {
    (0, _classCallCheck3.default)(this, ShowAssertionRetriesStatusCommand);

    this.type = _type2.default.showAssertionRetriesStatus;
    this.timeout = timeout;
};

var HideAssertionRetriesStatusCommand = exports.HideAssertionRetriesStatusCommand = function HideAssertionRetriesStatusCommand(success) {
    (0, _classCallCheck3.default)(this, HideAssertionRetriesStatusCommand);

    this.type = _type2.default.hideAssertionRetriesStatus;
    this.success = success;
};

var SetBreakpointCommand = exports.SetBreakpointCommand = function SetBreakpointCommand() {
    (0, _classCallCheck3.default)(this, SetBreakpointCommand);

    this.type = _type2.default.setBreakpoint;
};

var TestDoneCommand = exports.TestDoneCommand = function TestDoneCommand() {
    (0, _classCallCheck3.default)(this, TestDoneCommand);

    this.type = _type2.default.testDone;
};
