// testcafe/lib/test-run/commands/utils.js
'use strict';

exports.__esModule = true;
exports.isCommandRejectableByPageError = isCommandRejectableByPageError;
exports.canSetDebuggerBreakpointBeforeCommand = canSetDebuggerBreakpointBeforeCommand;
exports.isBrowserManipulationCommand = isBrowserManipulationCommand;
exports.isServiceCommand = isServiceCommand;
exports.isExecutableInTopWindowOnly = isExecutableInTopWindowOnly;

var _type = require('./type');

var _type2 = _interopRequireDefault(_type);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function isCommandRejectableByPageError(command) {
    return !isObservationCommand(command) && !isBrowserManipulationCommand(command) && !isServiceCommand(command) || isRejectablePrepareBrowserManipulationCommand(command) && !isWindowSwitchingCommand(command);
} // -------------------------------------------------------------
// WARNING: this file is used by both the client and the server.
// Do not use any browser or node-specific API!
// -------------------------------------------------------------


function isClientFunctionCommand(command) {
    return command.type === _type2.default.executeClientFunction || command.type === _type2.default.executeSelector;
}

function isObservationCommand(command) {
    return isClientFunctionCommand(command) || command.type === _type2.default.wait || command.type === _type2.default.assertion;
}

function isWindowSwitchingCommand(command) {
    return command.type === _type2.default.switchToIframe || command.type === _type2.default.switchToMainWindow;
}

function canSetDebuggerBreakpointBeforeCommand(command) {
    return command.type !== _type2.default.debug && !isClientFunctionCommand(command) && !isBrowserManipulationCommand(command) && !isServiceCommand(command);
}

function isBrowserManipulationCommand(command) {
    return command.type === _type2.default.takeScreenshot || command.type === _type2.default.takeScreenshotOnFail || command.type === _type2.default.resizeWindow || command.type === _type2.default.resizeWindowToFitDevice || command.type === _type2.default.maximizeWindow;
}

function isRejectablePrepareBrowserManipulationCommand(command) {
    return command.type === _type2.default.prepareBrowserManipulation && (command.manipulationCommandType === _type2.default.resizeWindow || command.manipulationCommandType === _type2.default.resizeWindowToFitDevice || command.manipulationCommandType === _type2.default.maximizeWindow);
}

function isServicePrepareBrowserManipulationCommand(command) {
    return command.type === _type2.default.prepareBrowserManipulation && command.manipulationCommandType === _type2.default.takeScreenshotOnFail;
}

function isServiceCommand(command) {
    return command.type === _type2.default.testDone || command.type === _type2.default.takeScreenshotOnFail || command.type === _type2.default.showAssertionRetriesStatus || command.type === _type2.default.hideAssertionRetriesStatus || command.type === _type2.default.setBreakpoint || isServicePrepareBrowserManipulationCommand(command);
}

function isExecutableInTopWindowOnly(command) {
    return command.type === _type2.default.testDone || command.type === _type2.default.prepareBrowserManipulation || command.type === _type2.default.switchToMainWindow || command.type === _type2.default.setNativeDialogHandler || command.type === _type2.default.getNativeDialogHistory || command.type === _type2.default.setTestSpeed || command.type === _type2.default.showAssertionRetriesStatus || command.type === _type2.default.hideAssertionRetriesStatus || command.type === _type2.default.setBreakpoint;
}
