// testcafe-hammerhead/lib/utils/parse5.js
'use strict';

exports.__esModule = true;
exports.createElement = createElement;
exports.insertElement = insertElement;
exports.removeNode = removeNode;
exports.findElementsByTagNames = findElementsByTagNames;
exports.walkElements = walkElements;
exports.createTextNode = createTextNode;
exports.removeAttr = removeAttr;
exports.getAttr = getAttr;
exports.setAttr = setAttr;

var _namespaces = require('../processing/dom/namespaces');

var ATTR_NAMESPACE_LOCAL_NAME_SEPARATOR = ':';

function getAttrName(attr) {
    return attr.prefix ? attr.prefix + ATTR_NAMESPACE_LOCAL_NAME_SEPARATOR + attr.name : attr.name;
}

function parseAttrName(attr) {
    var parts = attr.split(ATTR_NAMESPACE_LOCAL_NAME_SEPARATOR);

    if (parts.length === 2) {
        return {
            prefix: parts[0],
            name: parts[1]
        };
    }

    return {
        name: parts[0]
    };
}

function findAttr(el, name) {
    for (var i = 0; i < el.attrs.length; i++) {
        if (getAttrName(el.attrs[i]) === name) return el.attrs[i];
    }
    return null;
}

function createElement(tagName, attrs) {
    return {
        nodeName: tagName,
        tagName: tagName,
        attrs: attrs,
        childNodes: []
    };
}

function insertElement(el, parent) {
    el.namespaceURI = parent.namespaceURI;
    el.parentNode = parent;
    parent.childNodes.unshift(el);
}

function removeNode(node) {
    var parent = node.parentNode;
    var elIndex = parent.childNodes.indexOf(node);

    parent.childNodes.splice(elIndex, 1);
}

function findElementsByTagNames(root, tagNames) {
    var elements = {};

    walkElements(root, function (el) {
        if (tagNames.indexOf(el.tagName) !== -1) {
            elements[el.tagName] = elements[el.tagName] || [];
            elements[el.tagName].push(el);
        }
    });

    return elements;
}

function walkElements(el, processor) {
    if (el.nodeName !== '#document' && el.nodeName !== '#text' && el.nodeName !== '#documentType' && el.nodeName !== '#comment' && el.nodeName !== '#document-fragment') processor(el);

    if (el.childNodes) el.childNodes.forEach(function (child) {
        return walkElements(child, processor);
    });
}

function createTextNode(content, parent) {
    return {
        nodeName: '#text',
        value: content,
        parentNode: parent
    };
}

function removeAttr(el, name) {
    for (var i = 0; i < el.attrs.length; i++) {
        if (getAttrName(el.attrs[i]) === name) {
            el.attrs.splice(i, 1);

            return;
        }
    }
}

function getAttr(el, name) {
    var attr = findAttr(el, name);

    return attr ? attr.value : null;
}

function setAttr(el, name, value) {
    var attr = findAttr(el, name);

    if (attr) {
        attr.value = value;

        return value;
    }

    var parsedAttrName = parseAttrName(name);
    var newAttr = { name: parsedAttrName.name, value: value };

    if (parsedAttrName.prefix && _namespaces.NAMESPACE_PREFIX_MAP[parsedAttrName.prefix]) newAttr.namespace = _namespaces.NAMESPACE_PREFIX_MAP[parsedAttrName.prefix];

    el.attrs.push(newAttr);

    return value;
}
