// testcafe-browser-tools/lib/api/maximize.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator').default;

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default').default;

exports.__esModule = true;

var _osFamily = require('os-family');

var _osFamily2 = _interopRequireDefault(_osFamily);

var _findWindow = require('./find-window');

var _findWindow2 = _interopRequireDefault(_findWindow);

var _utilsExec = require('../utils/exec');

var _binaries = require('../binaries');

var _binaries2 = _interopRequireDefault(_binaries);

function getBoundsFromString(boundsString) {
    return boundsString.replace(/\n/g, '').split(', ');
}

function getWindowBounds(windowId, bundleId) {
    var boundsString;
    return _regeneratorRuntime.async(function getWindowBounds$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(_utilsExec.execFile(_binaries2.default.getWindowBounds, [windowId, bundleId]));

            case 2:
                boundsString = context$1$0.sent;
                return context$1$0.abrupt('return', getBoundsFromString(boundsString));

            case 4:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
}

function getWindowMaxBounds(windowId, bundleId) {
    var windowBounds, maxBoundsString;
    return _regeneratorRuntime.async(function getWindowMaxBounds$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(getWindowBounds(windowId, bundleId));

            case 2:
                windowBounds = context$1$0.sent;
                context$1$0.next = 5;
                return _regeneratorRuntime.awrap(_utilsExec.execFile(_binaries2.default.getWindowMaxBounds, windowBounds));

            case 5:
                maxBoundsString = context$1$0.sent;
                return context$1$0.abrupt('return', getBoundsFromString(maxBoundsString));

            case 7:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
}

function maximizeWindowMac(windowDescription) {
    var windowId, bundleId, windowBounds;
    return _regeneratorRuntime.async(function maximizeWindowMac$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                windowId = windowDescription.windowId;
                bundleId = windowDescription.bundleId;
                context$1$0.next = 4;
                return _regeneratorRuntime.awrap(getWindowMaxBounds(windowId, bundleId));

            case 4:
                windowBounds = context$1$0.sent;
                context$1$0.next = 7;
                return _regeneratorRuntime.awrap(_utilsExec.execFile(_binaries2.default.setWindowBounds, [windowId, bundleId].concat(windowBounds)));

            case 7:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
}
/**
 * Maximizes the specified browser window.
 * @function
 * @async
 * @name maximize
 * @param {string} pageTitle - The title of the web page opened in the window that should be maximized.
 **/

exports.default = function callee$0$0(pageTitle) {
    var windowDescription, commandArguments;
    return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
        while (1) switch (context$1$0.prev = context$1$0.next) {
            case 0:
                context$1$0.next = 2;
                return _regeneratorRuntime.awrap(_findWindow2.default(pageTitle));

            case 2:
                windowDescription = context$1$0.sent;

                if (windowDescription) {
                    context$1$0.next = 5;
                    break;
                }

                return context$1$0.abrupt('return');

            case 5:
                commandArguments = void 0;

                if (!_osFamily2.default.win) {
                    context$1$0.next = 10;
                    break;
                }

                commandArguments = [windowDescription.hwnd];
                context$1$0.next = 21;
                break;

            case 10:
                if (!_osFamily2.default.linux) {
                    context$1$0.next = 14;
                    break;
                }

                commandArguments = [windowDescription.windowId];
                context$1$0.next = 21;
                break;

            case 14:
                if (!_osFamily2.default.mac) {
                    context$1$0.next = 20;
                    break;
                }

                context$1$0.next = 17;
                return _regeneratorRuntime.awrap(maximizeWindowMac(windowDescription));

            case 17:
                return context$1$0.abrupt('return');

            case 20:
                return context$1$0.abrupt('return');

            case 21:
                context$1$0.next = 23;
                return _regeneratorRuntime.awrap(_utilsExec.execFile(_binaries2.default.maximize, commandArguments));

            case 23:
            case 'end':
                return context$1$0.stop();
        }
    }, null, this);
};

module.exports = exports.default;
