// testcafe/lib/browser/provider/index.js
'use strict';

exports.__esModule = true;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _pinkie = require('pinkie');

var _pinkie2 = _interopRequireDefault(_pinkie);

var _testcafeBrowserTools = require('testcafe-browser-tools');

var _testcafeBrowserTools2 = _interopRequireDefault(_testcafeBrowserTools);

var _osFamily = require('os-family');

var _osFamily2 = _interopRequireDefault(_osFamily);

var _delay = require('../../utils/delay');

var _delay2 = _interopRequireDefault(_delay);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*eslint-disable no-undef*/
function getTitle() {
    return document.title;
}

function getWindowDimensionsInfo() {
    return {
        width: window.innerWidth,
        height: window.innerHeight,
        outerWidth: window.outerWidth,
        outerHeight: window.outerHeight,
        availableWidth: screen.availWidth,
        availableHeight: screen.availHeight
    };
}
/*eslint-disable no-undef*/

var GET_TITLE_SCRIPT = getTitle.toString();
var GET_WINDOW_DIMENSIONS_INFO_SCRIPT = getWindowDimensionsInfo.toString();

var BROWSER_OPENING_DELAY = 2000;

var RESIZE_DIFF_SIZE = {
    width: 100,
    height: 100
};

function sumSizes(sizeA, sizeB) {
    return {
        width: sizeA.width + sizeB.width,
        height: sizeA.height + sizeB.height
    };
}

function subtractSizes(sizeA, sizeB) {
    return {
        width: sizeA.width - sizeB.width,
        height: sizeA.height - sizeB.height
    };
}

var BrowserProvider = function () {
    function BrowserProvider(plugin) {
        (0, _classCallCheck3.default)(this, BrowserProvider);

        this.plugin = plugin;
        this.initPromise = _pinkie2.default.resolve(false);

        this.isMultiBrowser = this.plugin.isMultiBrowser;
        // HACK: The browser window has different border sizes in normal and maximized modes. So, we need to be sure that the window is
        // not maximized before resizing it in order to keep the mechanism of correcting the client area size working. When browser is started,
        // we are resizing it for the first time to switch the window to normal mode, and for the second time - to restore the client area size.
        this.resizeCorrections = {};
        this.maxScreenSizes = {};
    }

    BrowserProvider.prototype._calculateResizeCorrections = function () {
        var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee(browserId) {
            var title, currentSize, etalonSize, resizedSize, correctionSize;
            return _regenerator2.default.wrap(function _callee$(_context) {
                while (1) {
                    switch (_context.prev = _context.next) {
                        case 0:
                            _context.next = 2;
                            return this.plugin.waitForConnectionReady(browserId);

                        case 2:
                            _context.next = 4;
                            return (0, _delay2.default)(BROWSER_OPENING_DELAY);

                        case 4:
                            _context.next = 6;
                            return this.plugin.runInitScript(browserId, GET_TITLE_SCRIPT);

                        case 6:
                            title = _context.sent;
                            _context.next = 9;
                            return _testcafeBrowserTools2.default.isMaximized(title);

                        case 9:
                            if (_context.sent) {
                                _context.next = 11;
                                break;
                            }

                            return _context.abrupt('return');

                        case 11:
                            _context.next = 13;
                            return this.plugin.runInitScript(browserId, GET_WINDOW_DIMENSIONS_INFO_SCRIPT);

                        case 13:
                            currentSize = _context.sent;
                            etalonSize = subtractSizes(currentSize, RESIZE_DIFF_SIZE);
                            _context.next = 17;
                            return _testcafeBrowserTools2.default.resize(title, currentSize.width, currentSize.height, etalonSize.width, etalonSize.height);

                        case 17:
                            _context.next = 19;
                            return this.plugin.runInitScript(browserId, GET_WINDOW_DIMENSIONS_INFO_SCRIPT);

                        case 19:
                            resizedSize = _context.sent;
                            correctionSize = subtractSizes(resizedSize, etalonSize);
                            _context.next = 23;
                            return _testcafeBrowserTools2.default.resize(title, resizedSize.width, resizedSize.height, etalonSize.width, etalonSize.height);

                        case 23:
                            _context.next = 25;
                            return this.plugin.runInitScript(browserId, GET_WINDOW_DIMENSIONS_INFO_SCRIPT);

                        case 25:
                            resizedSize = _context.sent;


                            correctionSize = sumSizes(correctionSize, subtractSizes(resizedSize, etalonSize));

                            this.resizeCorrections[browserId] = correctionSize;

                            _context.next = 30;
                            return _testcafeBrowserTools2.default.maximize(title);

                        case 30:
                        case 'end':
                            return _context.stop();
                    }
                }
            }, _callee, this);
        }));

        function _calculateResizeCorrections(_x) {
            return _ref.apply(this, arguments);
        }

        return _calculateResizeCorrections;
    }();

    BrowserProvider.prototype._calculateMacSizeLimits = function () {
        var _ref2 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee2(browserId) {
            var sizeInfo;
            return _regenerator2.default.wrap(function _callee2$(_context2) {
                while (1) {
                    switch (_context2.prev = _context2.next) {
                        case 0:
                            _context2.next = 2;
                            return this.plugin.waitForConnectionReady(browserId);

                        case 2:
                            _context2.next = 4;
                            return (0, _delay2.default)(BROWSER_OPENING_DELAY);

                        case 4:
                            _context2.next = 6;
                            return this.plugin.runInitScript(browserId, GET_WINDOW_DIMENSIONS_INFO_SCRIPT);

                        case 6:
                            sizeInfo = _context2.sent;


                            this.maxScreenSizes[browserId] = {
                                width: sizeInfo.availableWidth - (sizeInfo.outerWidth - sizeInfo.width),
                                height: sizeInfo.availableHeight - (sizeInfo.outerHeight - sizeInfo.height)
                            };

                        case 8:
                        case 'end':
                            return _context2.stop();
                    }
                }
            }, _callee2, this);
        }));

        function _calculateMacSizeLimits(_x2) {
            return _ref2.apply(this, arguments);
        }

        return _calculateMacSizeLimits;
    }();

    BrowserProvider.prototype._onOpenBrowser = function () {
        var _ref3 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee3(browserId) {
            return _regenerator2.default.wrap(function _callee3$(_context3) {
                while (1) {
                    switch (_context3.prev = _context3.next) {
                        case 0:
                            if (!_osFamily2.default.win) {
                                _context3.next = 5;
                                break;
                            }

                            _context3.next = 3;
                            return this._calculateResizeCorrections(browserId);

                        case 3:
                            _context3.next = 8;
                            break;

                        case 5:
                            if (!_osFamily2.default.mac) {
                                _context3.next = 8;
                                break;
                            }

                            _context3.next = 8;
                            return this._calculateMacSizeLimits(browserId);

                        case 8:
                        case 'end':
                            return _context3.stop();
                    }
                }
            }, _callee3, this);
        }));

        function _onOpenBrowser(_x3) {
            return _ref3.apply(this, arguments);
        }

        return _onOpenBrowser;
    }();

    BrowserProvider.prototype._closeLocalBrowser = function () {
        var _ref4 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee4(browserId) {
            return _regenerator2.default.wrap(function _callee4$(_context4) {
                while (1) {
                    switch (_context4.prev = _context4.next) {
                        case 0:
                            _context4.next = 2;
                            return _testcafeBrowserTools2.default.close(browserId);

                        case 2:
                        case 'end':
                            return _context4.stop();
                    }
                }
            }, _callee4, this);
        }));

        function _closeLocalBrowser(_x4) {
            return _ref4.apply(this, arguments);
        }

        return _closeLocalBrowser;
    }();

    BrowserProvider.prototype._resizeLocalBrowserWindow = function () {
        var _ref5 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee5(browserId, width, height, currentWidth, currentHeight) {
            return _regenerator2.default.wrap(function _callee5$(_context5) {
                while (1) {
                    switch (_context5.prev = _context5.next) {
                        case 0:
                            if (this.resizeCorrections[browserId]) {
                                width -= this.resizeCorrections[browserId].width;
                                height -= this.resizeCorrections[browserId].height;

                                delete this.resizeCorrections[browserId];
                            }

                            _context5.next = 3;
                            return _testcafeBrowserTools2.default.resize(browserId, currentWidth, currentHeight, width, height);

                        case 3:
                        case 'end':
                            return _context5.stop();
                    }
                }
            }, _callee5, this);
        }));

        function _resizeLocalBrowserWindow(_x5, _x6, _x7, _x8, _x9) {
            return _ref5.apply(this, arguments);
        }

        return _resizeLocalBrowserWindow;
    }();

    BrowserProvider.prototype._takeLocalBrowserScreenshot = function () {
        var _ref6 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee6(browserId, screenshotPath) {
            return _regenerator2.default.wrap(function _callee6$(_context6) {
                while (1) {
                    switch (_context6.prev = _context6.next) {
                        case 0:
                            _context6.next = 2;
                            return _testcafeBrowserTools2.default.screenshot(browserId, screenshotPath);

                        case 2:
                        case 'end':
                            return _context6.stop();
                    }
                }
            }, _callee6, this);
        }));

        function _takeLocalBrowserScreenshot(_x10, _x11) {
            return _ref6.apply(this, arguments);
        }

        return _takeLocalBrowserScreenshot;
    }();

    BrowserProvider.prototype._canResizeLocalBrowserWindowToDimensions = function () {
        var _ref7 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee7(browserId, width, height) {
            var maxScreenSize;
            return _regenerator2.default.wrap(function _callee7$(_context7) {
                while (1) {
                    switch (_context7.prev = _context7.next) {
                        case 0:
                            if (_osFamily2.default.mac) {
                                _context7.next = 2;
                                break;
                            }

                            return _context7.abrupt('return', true);

                        case 2:
                            maxScreenSize = this.maxScreenSizes[browserId];
                            return _context7.abrupt('return', width <= maxScreenSize.width && height <= maxScreenSize.height);

                        case 4:
                        case 'end':
                            return _context7.stop();
                    }
                }
            }, _callee7, this);
        }));

        function _canResizeLocalBrowserWindowToDimensions(_x12, _x13, _x14) {
            return _ref7.apply(this, arguments);
        }

        return _canResizeLocalBrowserWindowToDimensions;
    }();

    BrowserProvider.prototype._maximizeLocalBrowserWindow = function () {
        var _ref8 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee8(browserId) {
            return _regenerator2.default.wrap(function _callee8$(_context8) {
                while (1) {
                    switch (_context8.prev = _context8.next) {
                        case 0:
                            _context8.next = 2;
                            return _testcafeBrowserTools2.default.maximize(browserId);

                        case 2:
                        case 'end':
                            return _context8.stop();
                    }
                }
            }, _callee8, this);
        }));

        function _maximizeLocalBrowserWindow(_x15) {
            return _ref8.apply(this, arguments);
        }

        return _maximizeLocalBrowserWindow;
    }();

    BrowserProvider.prototype.init = function () {
        var _ref9 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee9() {
            var initialized;
            return _regenerator2.default.wrap(function _callee9$(_context9) {
                while (1) {
                    switch (_context9.prev = _context9.next) {
                        case 0:
                            _context9.next = 2;
                            return this.initPromise;

                        case 2:
                            initialized = _context9.sent;

                            if (!initialized) {
                                _context9.next = 5;
                                break;
                            }

                            return _context9.abrupt('return');

                        case 5:

                            this.initPromise = this.plugin.init().then(function () {
                                return true;
                            });

                            _context9.prev = 6;
                            _context9.next = 9;
                            return this.initPromise;

                        case 9:
                            _context9.next = 15;
                            break;

                        case 11:
                            _context9.prev = 11;
                            _context9.t0 = _context9['catch'](6);

                            this.initPromise = _pinkie2.default.resolve(false);

                            throw _context9.t0;

                        case 15:
                        case 'end':
                            return _context9.stop();
                    }
                }
            }, _callee9, this, [[6, 11]]);
        }));

        function init() {
            return _ref9.apply(this, arguments);
        }

        return init;
    }();

    BrowserProvider.prototype.dispose = function () {
        var _ref10 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee10() {
            var initialized;
            return _regenerator2.default.wrap(function _callee10$(_context10) {
                while (1) {
                    switch (_context10.prev = _context10.next) {
                        case 0:
                            _context10.next = 2;
                            return this.initPromise;

                        case 2:
                            initialized = _context10.sent;

                            if (initialized) {
                                _context10.next = 5;
                                break;
                            }

                            return _context10.abrupt('return');

                        case 5:

                            this.initPromise = this.plugin.dispose().then(function () {
                                return false;
                            });

                            _context10.prev = 6;
                            _context10.next = 9;
                            return this.initPromise;

                        case 9:
                            _context10.next = 15;
                            break;

                        case 11:
                            _context10.prev = 11;
                            _context10.t0 = _context10['catch'](6);

                            this.initPromise = _pinkie2.default.resolve(false);

                            throw _context10.t0;

                        case 15:
                        case 'end':
                            return _context10.stop();
                    }
                }
            }, _callee10, this, [[6, 11]]);
        }));

        function dispose() {
            return _ref10.apply(this, arguments);
        }

        return dispose;
    }();

    BrowserProvider.prototype.openBrowser = function () {
        var _ref11 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee11(browserId, pageUrl, browserName) {
            var isLocalBrowser;
            return _regenerator2.default.wrap(function _callee11$(_context11) {
                while (1) {
                    switch (_context11.prev = _context11.next) {
                        case 0:
                            _context11.next = 2;
                            return this.plugin.openBrowser(browserId, pageUrl, browserName);

                        case 2:
                            _context11.next = 4;
                            return this.plugin.isLocalBrowser(browserId);

                        case 4:
                            isLocalBrowser = _context11.sent;

                            if (!isLocalBrowser) {
                                _context11.next = 8;
                                break;
                            }

                            _context11.next = 8;
                            return this._onOpenBrowser(browserId);

                        case 8:
                        case 'end':
                            return _context11.stop();
                    }
                }
            }, _callee11, this);
        }));

        function openBrowser(_x16, _x17, _x18) {
            return _ref11.apply(this, arguments);
        }

        return openBrowser;
    }();

    BrowserProvider.prototype.closeBrowser = function () {
        var _ref12 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee12(browserId) {
            return _regenerator2.default.wrap(function _callee12$(_context12) {
                while (1) {
                    switch (_context12.prev = _context12.next) {
                        case 0:
                            _context12.next = 2;
                            return this.plugin.closeBrowser(browserId);

                        case 2:
                        case 'end':
                            return _context12.stop();
                    }
                }
            }, _callee12, this);
        }));

        function closeBrowser(_x19) {
            return _ref12.apply(this, arguments);
        }

        return closeBrowser;
    }();

    BrowserProvider.prototype.getBrowserList = function () {
        var _ref13 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee13() {
            return _regenerator2.default.wrap(function _callee13$(_context13) {
                while (1) {
                    switch (_context13.prev = _context13.next) {
                        case 0:
                            _context13.next = 2;
                            return this.plugin.getBrowserList();

                        case 2:
                            return _context13.abrupt('return', _context13.sent);

                        case 3:
                        case 'end':
                            return _context13.stop();
                    }
                }
            }, _callee13, this);
        }));

        function getBrowserList() {
            return _ref13.apply(this, arguments);
        }

        return getBrowserList;
    }();

    BrowserProvider.prototype.isValidBrowserName = function () {
        var _ref14 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee14(browserName) {
            return _regenerator2.default.wrap(function _callee14$(_context14) {
                while (1) {
                    switch (_context14.prev = _context14.next) {
                        case 0:
                            _context14.next = 2;
                            return this.plugin.isValidBrowserName(browserName);

                        case 2:
                            return _context14.abrupt('return', _context14.sent);

                        case 3:
                        case 'end':
                            return _context14.stop();
                    }
                }
            }, _callee14, this);
        }));

        function isValidBrowserName(_x20) {
            return _ref14.apply(this, arguments);
        }

        return isValidBrowserName;
    }();

    BrowserProvider.prototype.resizeWindow = function () {
        var _ref15 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee15(browserId, width, height, currentWidth, currentHeight) {
            var isLocalBrowser, supportedFeatures;
            return _regenerator2.default.wrap(function _callee15$(_context15) {
                while (1) {
                    switch (_context15.prev = _context15.next) {
                        case 0:
                            _context15.next = 2;
                            return this.plugin.isLocalBrowser(browserId);

                        case 2:
                            isLocalBrowser = _context15.sent;
                            _context15.next = 5;
                            return this.plugin.hasCustomActionForBrowser(browserId);

                        case 5:
                            supportedFeatures = _context15.sent;

                            if (!(isLocalBrowser && !supportedFeatures.hasResizeWindow)) {
                                _context15.next = 10;
                                break;
                            }

                            _context15.next = 9;
                            return this._resizeLocalBrowserWindow(browserId, width, height, currentWidth, currentHeight);

                        case 9:
                            return _context15.abrupt('return');

                        case 10:
                            _context15.next = 12;
                            return this.plugin.resizeWindow(browserId, width, height, currentWidth, currentHeight);

                        case 12:
                        case 'end':
                            return _context15.stop();
                    }
                }
            }, _callee15, this);
        }));

        function resizeWindow(_x21, _x22, _x23, _x24, _x25) {
            return _ref15.apply(this, arguments);
        }

        return resizeWindow;
    }();

    BrowserProvider.prototype.canResizeWindowToDimensions = function () {
        var _ref16 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee16(browserId, width, height) {
            var isLocalBrowser, supportedFeatures;
            return _regenerator2.default.wrap(function _callee16$(_context16) {
                while (1) {
                    switch (_context16.prev = _context16.next) {
                        case 0:
                            _context16.next = 2;
                            return this.plugin.isLocalBrowser(browserId);

                        case 2:
                            isLocalBrowser = _context16.sent;
                            _context16.next = 5;
                            return this.plugin.hasCustomActionForBrowser(browserId);

                        case 5:
                            supportedFeatures = _context16.sent;

                            if (!(isLocalBrowser && !supportedFeatures.hasCanResizeWindowToDimensions)) {
                                _context16.next = 10;
                                break;
                            }

                            _context16.next = 9;
                            return this._canResizeLocalBrowserWindowToDimensions(browserId, width, height);

                        case 9:
                            return _context16.abrupt('return', _context16.sent);

                        case 10:
                            _context16.next = 12;
                            return this.plugin.canResizeWindowToDimensions(browserId, width, height);

                        case 12:
                            return _context16.abrupt('return', _context16.sent);

                        case 13:
                        case 'end':
                            return _context16.stop();
                    }
                }
            }, _callee16, this);
        }));

        function canResizeWindowToDimensions(_x26, _x27, _x28) {
            return _ref16.apply(this, arguments);
        }

        return canResizeWindowToDimensions;
    }();

    BrowserProvider.prototype.maximizeWindow = function () {
        var _ref17 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee17(browserId) {
            var isLocalBrowser, supportedFeatures;
            return _regenerator2.default.wrap(function _callee17$(_context17) {
                while (1) {
                    switch (_context17.prev = _context17.next) {
                        case 0:
                            _context17.next = 2;
                            return this.plugin.isLocalBrowser(browserId);

                        case 2:
                            isLocalBrowser = _context17.sent;
                            _context17.next = 5;
                            return this.plugin.hasCustomActionForBrowser(browserId);

                        case 5:
                            supportedFeatures = _context17.sent;

                            if (!(isLocalBrowser && !supportedFeatures.hasCanResizeWindowToDimensions)) {
                                _context17.next = 10;
                                break;
                            }

                            _context17.next = 9;
                            return this._maximizeLocalBrowserWindow(browserId);

                        case 9:
                            return _context17.abrupt('return', _context17.sent);

                        case 10:
                            _context17.next = 12;
                            return this.plugin.maximizeWindow(browserId);

                        case 12:
                            return _context17.abrupt('return', _context17.sent);

                        case 13:
                        case 'end':
                            return _context17.stop();
                    }
                }
            }, _callee17, this);
        }));

        function maximizeWindow(_x29) {
            return _ref17.apply(this, arguments);
        }

        return maximizeWindow;
    }();

    BrowserProvider.prototype.takeScreenshot = function () {
        var _ref18 = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee18(browserId, screenshotPath, pageWidth, pageHeight) {
            var isLocalBrowser, supportedFeatures;
            return _regenerator2.default.wrap(function _callee18$(_context18) {
                while (1) {
                    switch (_context18.prev = _context18.next) {
                        case 0:
                            _context18.next = 2;
                            return this.plugin.isLocalBrowser(browserId);

                        case 2:
                            isLocalBrowser = _context18.sent;
                            _context18.next = 5;
                            return this.plugin.hasCustomActionForBrowser(browserId);

                        case 5:
                            supportedFeatures = _context18.sent;

                            if (!(isLocalBrowser && !supportedFeatures.hasTakeScreenshot)) {
                                _context18.next = 10;
                                break;
                            }

                            _context18.next = 9;
                            return this._takeLocalBrowserScreenshot(browserId, screenshotPath, pageWidth, pageHeight);

                        case 9:
                            return _context18.abrupt('return');

                        case 10:
                            _context18.next = 12;
                            return this.plugin.takeScreenshot(browserId, screenshotPath, pageWidth, pageHeight);

                        case 12:
                        case 'end':
                            return _context18.stop();
                    }
                }
            }, _callee18, this);
        }));

        function takeScreenshot(_x30, _x31, _x32, _x33) {
            return _ref18.apply(this, arguments);
        }

        return takeScreenshot;
    }();

    return BrowserProvider;
}();

exports.default = BrowserProvider;
module.exports = exports['default'];
